helm-tests/render_test.go

```diff
@@ -0,0 +1,118 @@
+package main_test
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+
+	appv1beta1 "k8s.io/api/apps/v1beta1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	polv1beta1 "k8s.io/api/policy/v1beta1"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+type AllResources struct {
+	calicoEtcdSecrets        *v1.Secret
+	calicoConfig             *v1.ConfigMap
+	calicoNode               *v1beta1.DaemonSet
+	calicoNodeServiceAccount *v1.ServiceAccount
+	calicoKubeControllers    *v1beta1.Deployment
+	calicoTypha              *appv1beta1.Deployment
+	kubeControllersSA        *v1.ServiceAccount
+	calicoTyphaService       *v1.Service
+	calicoTyphaPDB           *polv1beta1.PodDisruptionBudget
+}
+
+var (
+	chartPath string
+)
+
+func init() {
+	flag.StringVar(&chartPath, "chart-path", "../_includes/master/charts/calico", "path to the chart")
+}
+
+// TODO: Add call to kubeval to verify helm resources are valid
+func render(valuesyml string) (AllResources, error) {
+	f, err := ioutil.TempFile("", "helmfv")
+	if err != nil {
+		return AllResources{}, err
+	}
+
+	_, err = f.WriteString(valuesyml)
+	if err != nil {
+		return AllResources{}, err
+	}
+	if err := f.Close(); err != nil {
+		return AllResources{}, err
+	}
+
+	versionsFilePath := fmt.Sprintf("%s/%s", chartPath, "versions-values.yaml")
+	if _, err := os.Stat(versionsFilePath); err != nil && os.IsNotExist(err) {
+		return AllResources{}, fmt.Errorf("versions-values.yaml does not exist. Create one at %s.", versionsFilePath)
+	}
+
+	cmd := exec.Command("helm", "template", "-f", f.Name())
+	cmd.Args = append(cmd.Args, chartPath)
+	cmd.Args = append(cmd.Args, "-f", versionsFilePath)
+
+	var (
+		stdout bytes.Buffer
+		stderr bytes.Buffer
+	)
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+
+	if err = cmd.Run(); err != nil {
+		fmt.Println(stderr.String())
```

Comment:
Do we need to log _and_ return the error? I suggest just getting rid of the log line and have the calling code log the error if any is returned.